app/src/main/java/org/openobservatory/ooniprobe/activity/MeasurementDetailActivity.java

```diff
@@ -193,13 +199,39 @@ public boolean onCreateOptionsMenu(Menu menu) {
     public boolean onOptionsItemSelected(MenuItem item) {
         switch (item.getItemId()) {
             case R.id.rawData:
+                //Try to open file, if it doesn't exist dont show Error dialog immediately but try to download the json from internet
                 try {
                     File entryFile = Measurement.getEntryFile(this, measurement.id, measurement.test_name);
                     String json = FileUtils.readFileToString(entryFile, Charset.forName("UTF-8"));
                     json = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create().toJson(new JsonParser().parse(json));
                     startActivity(TextActivity.newIntent(this, json));
                 } catch (Exception e) {
-                    e.printStackTrace();
+                    getApiClient().getMeasurement(measurement.report_id, null).enqueue(new GetMeasurementsCallback() {
+                        @Override
+                        public void onSuccess(ApiMeasurement.Result result) {
+                            new OkHttpClient().newCall(new Request.Builder().url(result.measurement_url).build()).enqueue(new GetMeasurementCallback() {
```

Comment:
It seems to me in other cases we have specific `OkHttpClient`. Also, IIRC, the documentation says one should not create more than one such client per app. So, I'm confused as why we need a new client here?

Reply:
One client per App or per Activity? Where did you read that?

Reply:
I've read the documentation, since the url to call arrives directly from backend and could change without warning, it makes sense to make a single instance each time. 
In this case I'm making a different download every time and could be any random url.
The singleton OkHttpClient makes sense when the calls are made to a unique backend. If you see it in the code there is already a OkHttpClient singleton for APIClient and for OrchestraClient

Reply:
Please, add a comment that we may want to better understand this. The documentation of OkHttp suggests that we should have a single OkHttp client instance. We're not doing that, so we should document why.

Reply:
It seems we want to have different OkHttp clients for different concerns:

```Java
    private OkHttpClient getOkHttpClient() {
        if (okHttpClient == null) {
            HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
            logging.level(BuildConfig.DEBUG ? HttpLoggingInterceptor.Level.BODY : HttpLoggingInterceptor.Level.BASIC);
            okHttpClient = new OkHttpClient.Builder().addInterceptor(logging).build();
        }
        return okHttpClient;
    }

    public OONIOrchestraClient getOrchestraClient() {
        if (orchestraClient == null) {
            orchestraClient = new Retrofit.Builder()
                    .baseUrl(BuildConfig.OONI_ORCHESTRATE_BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(getOkHttpClient())
                    .build().create(OONIOrchestraClient.class);
        }
        return orchestraClient;
    }

    public OONIAPIClient getApiClient() {
        if (apiClient == null) {
            apiClient = new Retrofit.Builder()
                    .baseUrl(BuildConfig.OONI_API_BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(getOkHttpClient())
                    .build().create(OONIAPIClient.class);
        }
        return apiClient;
    } 
```

The first of these three singletons is probably what we want to use here. (The rationale of concerns separation is that you don't want to mix factories and builders for different concerns, which seems reasonable to me.)